### ISSUE TITLE
Is there any way to set cookie domain per request rather than per server launch?

### ISSUE THREAD

**Issue by Kumagor0** [koajs/session#188]
Right now I have the following code in my index.js:

```
const Koa = require('koa');
const session = require('koa-session');

const app = new Koa();

app.use(
  session(
    {
      key: SSID,
      maxAge: SESSION_TTL,
      rolling: true,
      httpOnly: true,
      store: SessionStore,
      domain: DOMAIN,
    },
    app
  )
);
```

The problem is, it sets domain once and for all and then passes that value to ctx.cookie.set() every time. What I need is customizable cookie domain based on request values like hostname or query etc. I can elaborate on the use case I need that for, if that's of any relevance. I don't see how can I do that currently, but I found a solution which basically creates new koa-session middleware on every request:

```
app.use(async (ctx, next) => {
  const { cookieDomain } = ctx.query;

  await session(
    {
      key: SSID,
      maxAge: SESSION_TTL,
      rolling: true,
      httpOnly: true,
      store: SessionStore,
      domain: cookieDomain || DOMAIN,
    },
    app
  )(ctx, next);
});
```

This would work, but unfortunately 2 out of 3 properties defined [here](https://github.com/koajs/session/blob/master/index.js#L126) have configurable set to false implicitly, so on the second request I get "TypeError: Cannot redefine property: sessionOptions" error.

So, I have 3 questions:

1. Is there any way to achieve what I need with koa-session right now that I've missed?

2. If previous answer is "no", is there any real reason sessionOptions and [CONTEXT_SESSION] are non-configurable?

3. If previous answer is "no", can it be changed to configurable? What needs to be done for that (do I need to submit PR or something)?

**Comment by Natouriano** [comment]
@Kumagor0 .. I'm facing the same issue.. did you have any luck with resolving this?

**Comment by Kumagor0** [comment]
@Natouriano no, not yet. Fortunately that's not a pressing matter for me, but if it would, I'd just fork this repo, make the changes I proposed in original post, publish it under new name and use that package instead of koa-session.

**Comment by olso** [comment]
This seems to do the job https://github.com/Secbone/koa-session2

**Comment by Kumagor0** [comment]
@olso Does it? Because looking at the docs, 

```
app.use(session({
    key: "SESSIONID",   //default "koa:sess"
}));
```

looks like setting all options once and for all.

**Comment by olso** [comment]
![Screenshot 2019-12-27 at 17 39 55](https://user-images.githubusercontent.com/7975158/71524984-1a6ed080-28d0-11ea-82ca-ddca956e1191.png)

**Comment by olso** [comment]
What prevents you from wrapping it inside another function?

![Screenshot 2019-12-27 at 18 50 49](https://user-images.githubusercontent.com/7975158/71527030-d1bc1500-28d9-11ea-91ad-2badb4e4ce1b.png)

My issue with `koa-session` is that you have to pass the Koa instance into it.

**Comment by zachsa** [comment]
I find that I can achieve this like so:

```js
  .use(async (ctx, next) => {
    const protocol = ctx.protocol
    const isHttp = protocol === 'http'
    return koaSession(
      {
        key: 'koa.sess',
        maxAge: ...,
        autoCommit: true,
        overwrite: true,
        httpOnly: true,
        signed: true,
        rolling: true,
        renew: false,
        secure: isHttp ? false : true,
        sameSite: isHttp ? 'lax' : 'none',
      },
      publicApp
    )(ctx, next)
  })
```

koa-session2 wasn't a drop in replacement for me. 

It looks like these two lines of code should ideally not be run on every http request: https://github.com/koajs/session/blob/master/index.js#L34-L35

**Comment by rahulgi** [comment]
Created a draft PR to support modifying the cookie domain on a per-request basis - https://github.com/koajs/session/pull/227
